### PR TITLE
[SimplifyCFG] handle monotonic wrapped case for D150943

### PR DESCRIPTION
**Pull request by khei4**
This is a follow-up on the miscompile for wrapping the flag for the switch to the lookup table on SimplifyCFG. Checking the switch results' monotonicity isn't sufficient to say linear map calculation has no wrapping. Monotonicity is only necessary, and further checking whether wrapping happens on the maximum index is required to be sufficient.
For example( @mikaelholmen  gave ), https://alive2.llvm.org/ce/z/So_fzw
Total linear map `(-2, -1, 1, 2)  |→ (-4, -2, 0, 2)` is monotonic, but actual calculation contains wrapping `(-2, -1, 1, 2)  |→ (0, 1, 2, 3) (*2) |→ (0, 2, 4(wrapped), 6(wrapped)) (-4)|→ (-4, -2, 0, 2).`

Also on the Linear map calculation, checking the multiplication wrapping with monotonicity is enough.(If the addition is only wrapped not with multiplication, contradicts with monotonicity. )

### CODE REVIEW

File: llvm/test/Transforms/SimplifyCFG/X86/switch_to_lookup_table.ll
Comment (mikaelholmen) [inline llvm/test/Transforms/SimplifyCFG/X86/switch_to_lookup_table.ll:2043]: Not sure if you really want to leave those "; Assume 1" comments etc that I made in the testcase when you submit it?

File: llvm/test/Transforms/SimplifyCFG/X86/switch_to_lookup_table.ll
Comment (mikaelholmen) [inline llvm/test/Transforms/SimplifyCFG/X86/switch_to_lookup_table.ll:2058]: Same here

File: llvm/test/Transforms/SimplifyCFG/X86/switch_to_lookup_table.ll
Comment (mikaelholmen) [inline llvm/test/Transforms/SimplifyCFG/X86/switch_to_lookup_table.ll:2063]: and here

File: llvm/test/Transforms/SimplifyCFG/X86/switch_to_lookup_table.ll
Comment (mikaelholmen) [inline llvm/test/Transforms/SimplifyCFG/X86/switch_to_lookup_table.ll:2067]: and here

**Comment by mikaelholmen** [comment]
Thanks @khei4 !
This looks good to me and I've verified that it solves the original miscompile problem I saw but I'd prefer if @nikic has a look too.

**Comment by djkoloski** [comment]
We ran into this on Fuchsia, which builds using tip-of-tree rustc. Once this is merged, can we get it added to the `llvmorg-17.0.0-rc4` tag, or a new release candidate tag that I can pull the fix back up into rust?

**Comment by DianQK** [comment]
> We ran into this on Fuchsia, which builds using tip-of-tree rustc. Once this is merged, can we get it added to the `llvmorg-17.0.0-rc4` tag, or a new release candidate tag that I can pull the fix back up into rust?

Bug fixes are always welcome. If you want to understand the process, refer to https://m.youtube.com/watch?v=onaNb2U1Od8.

**Comment by nikic** [comment]
Can you please update the PR description with more details about what the problematic case is an how this fixes it?

My understanding is that in this case we are mapping (after index adjustment) the indices (0, 1, 2, 3) to (-4, -2, 0, 2) so we have a monotonic sequence, but this is done by doing a multiply first, so we go through (0, 2, -4, -2) first, which wraps.

File: llvm/lib/Transforms/Utils/SimplifyCFG.cpp
Comment (nikic) [inline llvm/lib/Transforms/Utils/SimplifyCFG.cpp:6066]: NMon -> NonMonotonic would be slightly clearer imho.

File: llvm/lib/Transforms/Utils/SimplifyCFG.cpp
Comment (nikic) [inline llvm/lib/Transforms/Utils/SimplifyCFG.cpp:6098]: You can directly use `(void)` on the return value, no need to store it into an unused variable.

File: llvm/lib/Transforms/Utils/SimplifyCFG.cpp
Comment (nikic) [inline llvm/lib/Transforms/Utils/SimplifyCFG.cpp:6095]: Is the sadd_ov check here needed as well? If yes, is it possible to add a test case that needs it?

File: llvm/lib/Transforms/Utils/SimplifyCFG.cpp
Comment (nikic) [inline llvm/lib/Transforms/Utils/SimplifyCFG.cpp:6094]: MayMWrap -> MayWrap or MayMulWrap

**Review by nikic — APPROVED** [review]
LGTM

**Comment by djkoloski** [comment]
I'm having trouble backporting this to 17.x. I opened #66373 to backport it but it got closed automatically because the 17.x branch does not have the latest changes which remove the repo lockdown. So I opened #66375 to disable the repo lockdown on the 17.x branch but that was also closed automatically (I didn't know that GHA would run using the in-tree actions instead of the ones in the PR). Any suggestions? Should I go through Phabricator for this?

**Comment by nikic** [comment]
@djkoloski Backports still follow the process at https://llvm.org/docs/GitHub.html#backporting-fixes-to-the-release-branches.

**Comment by DianQK** [comment]
@djkoloski You need to create an issue first. The presentation above has more details.
I can do this on the weekends if needed.

**Comment by djkoloski** [comment]
Thanks for the help everyone, sorry for the confusion. I've created issue #66396 so a release manager can decide whether this change should be added to the LLVM 17.0.X milestone.

**Pull request merged**
